lib/jekyll/renderer.rb

```diff
@@ -144,67 +133,113 @@ def render_liquid(content, payload, info, path = nil)
     #
     # layout - the layout to check
     #
-    # Returns true if the layout is invalid, false if otherwise
+    # Returns Boolean true if the layout is invalid, false if otherwise
     def invalid_layout?(layout)
       !document.data["layout"].nil? && layout.nil? && !(document.is_a? Jekyll::Excerpt)
     end
 
-    # Render layouts and place given content inside.
-    #
-    # content - the content to be placed in the layout
-    #
+    # Render layouts and place document content inside.
     #
-    # Returns the content placed in the Liquid-rendered layouts
+    # Returns nothing
     def place_in_layouts(content, payload, info)
-      output = content.dup
+      @output = content.dup
       layout = layouts[document.data["layout"]]
-
-      Jekyll.logger.warn(
-        "Build Warning:",
-        "Layout '#{document.data["layout"]}' requested in "\
-        "#{document.relative_path} does not exist."
-      ) if invalid_layout? layout
+      validate_layout(layout)
 
       used = Set.new([layout])
 
       # Reset the payload layout data to ensure it starts fresh for each page.
       payload["layout"] = nil
 
       while layout
-        payload["content"] = output
-        payload["layout"]  = Utils.deep_merge_hashes(layout.data, payload["layout"] || {})
-
-        output = render_liquid(
-          layout.content,
-          payload,
-          info,
-          layout.relative_path
-        )
-
-        # Add layout to dependency tree
-        site.regenerator.add_dependency(
-          site.in_source_dir(document.path),
-          site.in_source_dir(layout.path)
-        ) if document.write?
-
-        if (layout = layouts[layout.data["layout"]])
+        render_layout(layout, info)
+        add_layout_to_dependency(layout)
```

Comment:
Let's rephrase this method – I was a bit confused by its name. What do you think about `add_regenerator_dependencies(layout)`?